### ISSUE TITLE
Moq1100 false positive: GetRefKind reads only Modifiers[0], so `scoped ref` (and `ref readonly`) lambda parameters are misclassified

### ISSUE THREAD

**Issue by rjmurillo** [rjmurillo/moq.analyzers#1262]
## Summary

An internal code audit found that `CallbackSignatureShouldMatchMockedMethodAnalyzer` (rule **Moq1100**) determines a callback lambda parameter's ref kind by inspecting **only the first modifier token** (`src/Analyzers/CallbackSignatureShouldMatchMockedMethodAnalyzer.cs:244-258`, switch at `:251`). For a parameter written `scoped ref string data`, `Modifiers[0]` is `scoped`, the switch falls through to `RefKind.None`, and the analyzer reports a spurious Moq1100 on a callback whose signature actually matches the mocked method. The same single-token read also mishandles C# 12 `ref readonly` parameters (two tokens: `ref`, `readonly`).

All premises were verified against the pinned toolchain (facts below). The fix is to stop re-deriving the ref kind from syntax tokens at all: the method already resolves the lambda parameter's `IParameterSymbol` two lines earlier, and the symbol's `RefKind` is authoritative for every modifier combination and ordering.

## Verified facts

Verified by compiling against the pinned Microsoft.CodeAnalysis 4.8.0 packages:

- `scoped` is legal on lambda parameters (with explicit type) from **C# 11**; `ref readonly` parameters require **C# 12**. Both compile cleanly under Roslyn 4.8 with `LanguageVersion.CSharp12`.
- The repo builds with `<LangVersion>default</LangVersion>` (`build/targets/compiler/Compiler.props:5`), and the test project compiles test sources through Microsoft.CodeAnalysis.CSharp **4.8** (`Directory.Packages.props:21-22`) with the testing framework's default language version — C# 12 for Roslyn 4.8 — against `ReferenceAssemblies.Net.Net80` (`tests/Moq.Analyzers.Test/Helpers/ReferenceAssemblyCatalog.cs:48-56`). So `scoped ref`/`ref readonly` test sources compile in the test harness.
- For a lambda parameter declared `(scoped ref int x)`: `parameter.Modifiers` is `[scoped, ref]` (so `Modifiers[0].ValueText == "scoped"`), and `SemanticModel.GetDeclaredSymbol(parameter).RefKind == RefKind.Ref`.
- `RefKind.RefReadOnlyParameter` **exists in Microsoft.CodeAnalysis 4.8.0** (confirmed in the 4.8.0 package's `Microsoft.CodeAnalysis.xml`: "Indicates a \"ref readonly\" parameter"), and a method parameter declared `ref readonly int y` has `RefKind == RefKind.RefReadOnlyParameter`. No fallback mapping to `RefKind.In` is needed — comparing symbol `RefKind` to symbol `RefKind` handles it natively (a mocked `ref readonly` parameter matches a `ref readonly` lambda parameter, and correctly mismatches `ref`/`in`).
- Moq relevance: an interface method may declare `scoped ref` parameters; `mock.Setup(m => m.M(ref It.Ref<T>.IsAny)).Callback((scoped ref T x) => { })` is legitimate user code. Today it is flagged because the analyzer computes `RefKind.None` for the lambda parameter while the mocked parameter is `RefKind.Ref`.

## Current behavior

`src/Analyzers/CallbackSignatureShouldMatchMockedMethodAnalyzer.cs:218-258`:

```csharp
    private static bool ParameterTypesMatch(
        SemanticModel semanticModel,
        IParameterSymbol mockedParameter,
        ParameterSyntax lambdaParameter,
        CancellationToken cancellationToken)
    {
        // Use semantic model to resolve the parameter type, handling both
        // explicitly typed and implicitly typed lambda parameters.
        IParameterSymbol? parameterSymbol = semanticModel.GetDeclaredSymbol(lambdaParameter, cancellationToken);
        ITypeSymbol? lambdaParameterTypeSymbol = parameterSymbol?.Type;

        if (lambdaParameterTypeSymbol is null || lambdaParameterTypeSymbol.TypeKind == TypeKind.Error)
        {
            return false; // Type could not be resolved; treat as mismatch to avoid suppressing Moq1100
        }

        ITypeSymbol mockedParameterType = mockedParameter.Type;

        if (!semanticModel.HasConversion(mockedParameterType, lambdaParameterTypeSymbol))
        {
            return false;
        }

        return mockedParameter.RefKind == GetRefKind(lambdaParameter);
    }

    private static RefKind GetRefKind(ParameterSyntax parameter)
    {
        if (parameter.Modifiers.Count == 0)
        {
            return RefKind.None;
        }

        return parameter.Modifiers[0].ValueText switch
        {
            "ref" => RefKind.Ref,
            "out" => RefKind.Out,
            "in" => RefKind.In,
            _ => RefKind.None,
        };
    }
```

Concrete failure: mocked interface `int DoScopedRef(scoped ref string data);`, user callback `.Callback((scoped ref string data) => { })` — mocked `RefKind.Ref` vs computed `RefKind.None` → spurious Moq1100.

## Desired behavior

The ref-kind comparison uses the resolved lambda parameter symbol's `RefKind`, which Roslyn computes correctly for `ref`, `out`, `in`, `ref readonly`, `scoped ref`, `params`, and any future modifier combination or token ordering. Matching signatures (including `scoped ref` and `ref readonly`) produce no diagnostic; genuinely mismatched ref kinds still do.

Rationale for symbol-over-syntax (per the audit's proposed "iterate all modifiers" fix, superseded): iterating modifier tokens and hand-mapping `ref`+`readonly` → `RefKind.RefReadOnlyParameter`, skipping `scoped`, etc., re-implements what `GetDeclaredSymbol(...).RefKind` already returns, and would need maintenance for every future modifier. The symbol is already resolved on the line above (non-null is guaranteed past the type-null check), so the symbol-based fix is both simpler and authoritative — consistent with ADR-001's preference for semantic over textual decisions.

## Implementation plan

1. In `src/Analyzers/CallbackSignatureShouldMatchMockedMethodAnalyzer.cs`, replace `ParameterTypesMatch` with:

```csharp
    private static bool ParameterTypesMatch(
        SemanticModel semanticModel,
        IParameterSymbol mockedParameter,
        ParameterSyntax lambdaParameter,
        CancellationToken cancellationToken)
    {
        // Use semantic model to resolve the parameter type, handling both
        // explicitly typed and implicitly typed lambda parameters.
        IParameterSymbol? parameterSymbol = semanticModel.GetDeclaredSymbol(lambdaParameter, cancellationToken);
        ITypeSymbol? lambdaParameterTypeSymbol = parameterSymbol?.Type;

        if (parameterSymbol is null || lambdaParameterTypeSymbol is null || lambdaParameterTypeSymbol.TypeKind == TypeKind.Error)
        {
            return false; // Type could not be resolved; treat as mismatch to avoid suppressing Moq1100
        }

        ITypeSymbol mockedParameterType = mockedParameter.Type;

        if (!semanticModel.HasConversion(mockedParameterType, lambdaParameterTypeSymbol))
        {
            return false;
        }

        // Compare symbol RefKinds rather than re-deriving the lambda parameter's ref kind from
        // syntax tokens. The symbol handles all modifier combinations and orderings, including
        // `scoped ref` (RefKind.Ref) and C# 12 `ref readonly` (RefKind.RefReadOnlyParameter,
        // available since Microsoft.CodeAnalysis 4.8), which a Modifiers[0]-only read misreads
        // as RefKind.None.
        return mockedParameter.RefKind == parameterSymbol.RefKind;
    }
```

2. Delete `GetRefKind` (`:244-258`) — after step 1 it has zero callers (it is `private` and referenced only from `:241`; grep-verified).

3. No other members change. Note the code-fix side is already symbol-based: `CallbackSignatureShouldMatchMockedMethodFixer.BuildParameterList` generates modifiers from `parameterSymbol.RefKind` (`src/CodeFixes/CallbackSignatureShouldMatchMockedMethodFixer.cs:184`), so a fixed positive case on a `scoped ref` mocked parameter emits `(ref string data)` (without `scoped`), which compiles and is accepted by Moq.

## Tests

Extend `tests/Moq.Analyzers.Test/RefOutCallbackTests.cs`, mirroring its existing structure exactly (code-fix rows via `CodeFixVerifier<CallbackSignatureShouldMatchMockedMethodAnalyzer, CallbackSignatureShouldMatchMockedMethodFixer>`, `[Theory]` + `[MemberData(nameof(TestData))]`, `.WithNamespaces().WithMoqReferenceAssemblyGroups()`, shared `IFoo` template). Two changes:

1. Add a `scoped ref` method to the template interface (current template at `:26-32`):

```csharp
            internal interface IFoo
            {
                int DoRef(ref string data);
                bool DoOut(out int result);
                string DoIn(in DateTime timestamp);
                int DoScopedRef(scoped ref string data);
            }
```

2. Add rows to `TestData()`:

```csharp
            // scoped ref: matching callback must NOT be flagged (regression for the Modifiers[0] bug).
            // original == fixed and no markup => negative row.
            [
                """new Mock<IFoo>().Setup(m => m.DoScopedRef(ref It.Ref<string>.IsAny)).Callback((scoped ref string data) => { });""",
                """new Mock<IFoo>().Setup(m => m.DoScopedRef(ref It.Ref<string>.IsAny)).Callback((scoped ref string data) => { });""",
            ],

            // scoped ref: by-value callback parameter is still a genuine mismatch (positive row).
            // The fixer regenerates modifiers from RefKind, so the fix emits `ref` (scoped is not
            // part of RefKind and is not required at the callback site).
            [
                """new Mock<IFoo>().Setup(m => m.DoScopedRef(ref It.Ref<string>.IsAny)).Callback(({|Moq1100:string data|}) => { });""",
                """new Mock<IFoo>().Setup(m => m.DoScopedRef(ref It.Ref<string>.IsAny)).Callback((ref string data) => { });""",
            ],
```

Implementation note for the test author: the existing rows pass ref-parameter lambdas directly to `.Callback(...)` and run against both Moq 4.8.2 and 4.18.4 reference-assembly groups, so keep the same direct-lambda form. If the C# natural-delegate-type conversion of a `scoped ref` lambda to Moq's `Callback(Delegate)` overload fails to compile in the harness, wrap the lambda in an explicit delegate instead — add `internal delegate int ScopedRefCallback(scoped ref string data);` to the template and use `.Callback(new ScopedRefCallback((scoped ref string data) => { }))`; the analyzer's `TryGetCallbackLambda` explicitly supports the delegate-constructor form (`src/Analyzers/CallbackSignatureShouldMatchMockedMethodAnalyzer.cs:112-121`). Verify with the negative row first: before the fix it must FAIL (spurious Moq1100), after the fix it must pass — that failing-first check proves the row reproduces the bug.

Optionally (recommended, cheap): a `ref readonly` row pair in the same file — mocked `string DoRefReadonly(ref readonly DateTime timestamp);` with matching `(ref readonly DateTime timestamp)` callback as a negative row. Note `ref readonly` needs C# 12; the harness default satisfies this (see Verified facts).

All existing rows in `RefOutCallbackTests.cs` and `CallbackSignatureShouldMatchMockedMethodAnalyzerTests.cs`/`CallbackSignatureShouldMatchMockedMethodCodeFixTests.cs` must pass unchanged — for plain `ref`/`out`/`in` parameters the symbol `RefKind` equals what `GetRefKind` returned, so behavior there is identical.

## Constraints

- Analyzer assemblies target **netstandard2.0**; Roslyn pinned to **Microsoft.CodeAnalysis 4.8** per ADR-003. `RefKind.RefReadOnlyParameter` is present in 4.8 (verified above); no newer API is used. **No new package dependencies.**
- **ADR-001** (name-based checks allowed only as fast-path pre-filters before an authoritative symbol check): this change goes the right direction — it *removes* a token-text-based decision (`Modifiers[0].ValueText` switch) in favor of the authoritative symbol; no name-based logic is added.
- **ADR-007**: this analyzer already uses `RegisterOperationAction` (`:47-49`); unchanged.
- Tests are mandatory and specified above.

## Validation

1. `dotnet format` (no diff).
2. `dotnet build` — **0 warnings**.
3. `dotnet test --settings ./build/targets/tests/test.runsettings` — all tests pass, including the new rows (and confirm the negative row fails on the pre-fix analyzer to prove reproduction).
4. Perf: `tests/Moq.Analyzers.Benchmarks/Moq1100CallbackBenchmarks.cs` exists; run the documented flow (`build/scripts/perf/README.md`; CONTRIBUTING.md "Performance Testing Guidelines", line 811): `./build/scripts/perf/PerfCore.ps1 -projects "tests/Moq.Analyzers.Benchmarks/Moq.Analyzers.Benchmarks.csproj" -diff`. Expect neutral (one symbol read replaces a token switch).

## Acceptance criteria

- [ ] `(scoped ref T x)` callback parameters matching a `scoped ref` mocked parameter produce no Moq1100.
- [ ] Genuine ref-kind mismatches on such methods still produce Moq1100 with the existing message/span, and the code fix emits compilable parameters.
- [ ] `GetRefKind` is deleted; no syntax-token ref-kind derivation remains in the analyzer.
- [ ] `ref readonly` parameters compare via `RefKind.RefReadOnlyParameter` (no `RefKind.In` conflation).
- [ ] All existing tests pass unchanged; build 0 warnings.

## Out of scope

- Teaching the code fix to emit `scoped`/`ref readonly` modifiers (fix currently emits plain `ref`/`out`/`in` from `RefKind`; output remains compilable).
- Audit A-2 (unguarded `Arguments[0]` reachable through this analyzer's fixer) — separate issue.
- The unused ref/out/in fixture methods noted by the audit's test-quality pass (`CallbackSignatureShouldMatchMockedMethodAnalyzerTests.cs:123-130`) — separate test-quality item.
